src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,385 @@
+{
+    if (m_results.m_update_policy == Results::UpdatePolicy::Never)
+        return;
+    else if (!m_results.has_changed() && !force_update)
+        return;
+
+    {
+        util::CheckedUniqueLock lock(m_results.m_mutex);
+        m_results.ensure_up_to_date();
+    }
+
+    calculate_sections();
+}
+
+// This method will run in the following scenarios:
+// - SectionedResults is freshly created from Results
+// - The underlying Table in the Results collection has changed
+void SectionedResults::calculate_sections()
+{
+    auto r = m_results.snapshot();
+    size_t size = r.size();
+    m_sections.clear();
+    m_row_to_index_path.clear();
+    std::map<Mixed, size_t> key_to_section_index;
+    m_row_to_index_path.resize(size);
+
+    for (size_t i = 0; i < size; ++i) {
+        Mixed key = m_callback(r.get_any(i), r.get_realm());
+        // Disallow links as section keys. It would be uncommon to use them to begin with
+        // and if the object acting as the key was deleted bad things would happen.
+        if (!key.is_null() && key.get_type() == type_Link) {
+            throw std::logic_error("Links are not supported as section keys.");
+        }
+        auto it = key_to_section_index.find(key);
+        if (it == key_to_section_index.end()) {
+            auto idx = m_sections.size();
+            Section section;
+            section.key = key;
+            section.index = idx;
+            section.indices.push_back(i);
+            section.hash = key.hash();
+            m_sections.push_back(section);
+            it = key_to_section_index.insert({key, m_sections.size() - 1}).first;
+            m_row_to_index_path[i] = {it->second, section.indices.size() - 1};
+        }
+        else {
+            auto& section = m_sections[it->second];
+            section.indices.push_back(i);
+            m_row_to_index_path[i] = {it->second, section.indices.size() - 1};
+        }
+    }
+}
+
+size_t SectionedResults::size()
+{
+    calculate_sections_if_required();
+    return m_sections.size();
+}
+
+ResultsSection SectionedResults::operator[](size_t idx)
+{
+    return ResultsSection(this, idx);
+}
+
+NotificationToken SectionedResults::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                              KeyPathArray key_path_array) &
+{
+    return m_results.add_notification_callback(SectionedResultsNotificationHandler(*this, std::move(callback)),
+                                               std::move(key_path_array));
+}
+
+NotificationToken SectionedResults::add_notification_callback_for_section(
+    size_t section_index, SectionedResultsNotificatonCallback callback, KeyPathArray key_path_array)
+{
+    return m_results.add_notification_callback(
+        SectionedResultsNotificationHandler(*this, std::move(callback), section_index), std::move(key_path_array));
+}
+
+realm::ThreadSafeReference SectionedResults::thread_safe_reference()
```

Comment:
I don't see how this creates a TSR to a SectionedResults?

Reply:
This just creates a TSR of the underlying results and passes it back, Cocoa requires it for observation on a non-main queue. Should we just remove this method and use the results TSR method instead?

Reply:
Yes, this seems actively confusing to me.